include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+        -> xmasked_value<xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>>                                 \
+    {                                                                                                            \
+        using value_type = xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>;                               \
+        return e1.has_value() && e2.has_value() ? masked_value(e1.value() OP e2.value()) : missing<value_type>();\
+    }                                                                                                            \
+                                                                                                                 \
+    template <class T1, class B1, class T2, class B2>                                                            \
+    inline auto operator OP(const xtl::xoptional<T1, B1>& e1, const xmasked_value<T2, B2>& e2) noexcept          \
+        -> xmasked_value<xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>>                                 \
+    {                                                                                                            \
+        using value_type = xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>;                               \
+        return e1.has_value() && e2.has_value() ? masked_value(e1.value() OP e2.value()) : missing<value_type>();\
+    }                                                                                                            \
+                                                                                                                 \
+    template <class T1, class B1, class T2, class B2>                                                            \
+    inline auto operator OP(const xmasked_value<T1, B1>& e1, const xmasked_value<T2, B2>& e2) noexcept           \
+        -> xmasked_value<xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>>                                 \
+    {                                                                                                            \
+        using value_type = xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>;                               \
+        return e1.has_value() && e2.has_value() ? masked_value(e1.value() OP e2.value()) : missing<value_type>();\
+    }                                                                                                            \
+                                                                                                                 \
+    template <class T1, class B1, class T2>                                                                      \
+    inline auto operator OP(const xmasked_value<T1, B1>& e1, const T2& e2) noexcept                              \
+        -> disable_xoptional_like<T2, xmasked_value<xt::promote_type_t<std::decay_t<T1>, std::decay_t<T2>>>>     \
```

Comment:
The condition for disabling should now be based on `T2` being an `xmasked_value` rather than an `xoptional`

Reply:
That's why I created the `disable_xoptional_like` which disables the function if `T2` is an `xoptional` or `xmasked_value`, maybe we need a better name than `disable_xoptional_like`

Reply:
My point is we should only disable on `T2` being a `xmasked_value`, the `xoptional` should not be considered.